mbedtls/src/pkcs12/mod.rs

```diff
@@ -520,6 +530,26 @@ pub struct Pfx {
     raw_data: Vec<u8>,
 }
 
+#[derive(Debug, Clone)]
+enum Pkcs12Pbes {
+    Pbes1 {
+        algo: ObjectIdentifier,
+        params: Pkcs12PbeParams,
+    },
+    Pbes2(Pkcs12Pbes2Params),
+}
+
+impl Pkcs12Pbes {
+    fn decrypt(&self, ciphertext: &[u8], passphrase: &Passphrase) -> Pkcs12Result<Vec<u8>> {
```

Comment:
Is it possible to pass a closure computing the passphrase instead of storing it raw in the `Passphrase` object?

Reply:
while it is possible, there is basically only 1 valid path for the password format for each variant there and `Password` just encapsulates this

Reply:
ah, and also it's not really possible, the closure argument would need to be propagated up to internal `decrypt_data` in `Pfx`, but this function still needs both variants at the same time, to pick the right one depending on which pbe algorithm the `SafeBag` will be encrypted with. We could default to raw password and perform the ucs16 transaltion on the fly where needed. but it's needed in definitely more places than raw one, and each such transform allocs

Reply:
We can also assume that all the passwords passed everywhere are in ucs16 and only transform back to the regular &str form when decoding pbes2. This will remove the `Password` struct and result in only one needless alloc when translating back. Then it'd be worth documenting that all the other places what password format is it expecting, as it was expressed by just &[u8] which doesn't bring this information

Reply:
I could also rename `raw` to `utf8` so that it's clearer

Reply:
I am not sure that we are talking about the same, but glad for this discussion. (The closure idea was to prompt the end user for a password with caller-defined code, not map to the correct format. Let us drop this idea in any case as out of scope, because this was not before either and perhaps not something to include in rust-mbedtls anyway).

Since we're not going that road, I'm fine with the `Passphrase` struct. The API is easy enough (`Passphrase::new(...)`). I agree with `utf8` instead of `raw`.

Last thing, perhaps it should be called `Pkcs12Passphrase` or similar (in the case where there are other passphrases needed elsewhere in the future).

Reply:
:+1: I'll also rename EncryptionScheme same way

Reply:
done